File: src/main/java/pe/nanamochi/banchus/repositories/ScoreRepository.java
package pe.nanamochi.banchus.repositories;

import java.util.List;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;
import pe.nanamochi.banchus.entities.BeatmapRankedStatus;
import pe.nanamochi.banchus.entities.CountryCode;
import pe.nanamochi.banchus.entities.Mode;
import pe.nanamochi.banchus.entities.SubmissionStatus;
import pe.nanamochi.banchus.entities.db.Beatmap;
import pe.nanamochi.banchus.entities.db.Score;
import pe.nanamochi.banchus.entities.db.User;

@Repository
public interface ScoreRepository extends JpaRepository<Score, Integer> {
  List<Score>
      findTop100ByUserAndModeAndSubmissionStatusInAndBeatmapStatusInOrderByPerformancePointsDesc(
          User user,
          Mode mode,
          List<SubmissionStatus> submissionStatuses,
          List<BeatmapRankedStatus> beatmapRankedStatuses);

  int countByUserAndModeAndSubmissionStatusInAndBeatmapStatusIn(
      User user,
      Mode mode,
      List<SubmissionStatus> submissionStatuses,
      List<BeatmapRankedStatus> beatmapRankedStatuses);

  Score findFirstByBeatmapAndUserAndSubmissionStatusOrderByPerformancePointsDesc(
      Beatmap beatmap, User user, SubmissionStatus submissionStatus);

  // sin mods
  List<Score> findTop50ByBeatmapAndModeAndSubmissionStatusAndUser_RestrictedFalseOrderByScoreDesc(
      Beatmap beatmap, Mode mode, SubmissionStatus submissionStatus);

  List<Score>
      findTop50ByBeatmapAndModeAndSubmissionStatusAndUser_RestrictedFalseAndUser_CountryOrderByScoreDesc(
          Beatmap beatmap, Mode mode, SubmissionStatus submissionStatus, CountryCode country);

  List<Score>
      findTop50ByBeatmapAndModeAndModsAndSubmissionStatusAndUser_RestrictedFalseOrderByScoreDesc(
          Beatmap beatmap, Mode mode, int mods, SubmissionStatus submissionStatus);

  List<Score>
      findTop50ByBeatmapAndModeAndModsAndSubmissionStatusAndUser_RestrictedFalseAndUser_CountryOrderByScoreDesc(
          Beatmap beatmap,
          Mode mode,
          int mods,
          SubmissionStatus submissionStatus,
          CountryCode country);
}

File: banchus-common/src/main/kotlin/pe/nanamochi/banchus/database/repository/ScoreRepository.kt
package pe.nanamochi.banchus.database.repository

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import pe.nanamochi.banchus.database.entity.Beatmap
import pe.nanamochi.banchus.database.entity.Score
import pe.nanamochi.banchus.database.entity.User
import pe.nanamochi.banchus.domain.enums.BeatmapRankedStatus
import pe.nanamochi.banchus.domain.enums.CountryCode
import pe.nanamochi.banchus.domain.enums.Mode
import pe.nanamochi.banchus.domain.enums.SubmissionStatus

@Repository
interface ScoreRepository : JpaRepository<Score, Int> {
    @Query(
        """
        SELECT s
        FROM Score s
        JOIN s.user u
        WHERE s.beatmap = :beatmap
          AND (:mode IS NULL OR s.mode = :mode)
          AND (:mods IS NULL OR s.mods = :mods)
          AND (:country IS NULL OR u.country = :country)
          AND s.submissionStatus = :status
          AND u.privileges >= 1
        ORDER BY s.score DESC
        """
    )
    fun fetchBeatmapLeaderboard(
        beatmap: Beatmap,
        mode: Mode?,
        mods: Int?,
        status: SubmissionStatus,
        country: CountryCode?,
    ): List<Score>

    fun findFirstByBeatmapAndUserAndSubmissionStatusOrderByPerformancePointsDesc(
        beatmap: Beatmap,
        user: User,
        submissionStatus: SubmissionStatus,
    ): Score?

    fun findTop100ByUserAndModeAndSubmissionStatusInAndBeatmapStatusInOrderByPerformancePointsDesc(
        user: User,
        mode: Mode,
        submissionStatuses: List<SubmissionStatus>,
        beatmapRankedStatuses: List<BeatmapRankedStatus>,
    ): List<Score>
}
